File: solutions/javascript/PathSumIII/PathSumIII.js
/**
 * Definition for a binary tree node.
 * function TreeNode(val) {
 *     this.val = val;
 *     this.left = this.right = null;
 * }
 */
/**
 * @param {TreeNode} root
 * @param {number} sum
 * @return {number}
 */
var pathSum = function(root, sum) {
    if (!root) {
        return 0;
    }
    return xxx(root, 0, sum) + pathSum(root.left, sum) + pathSum(root.right, sum);
};
var xxx = function(root, pre, sum) {
    if (!root) {
        return 0;
    }
    let cur_sum = pre + root.val;
    return (cur_sum == sum ? 1 : 0) + xxx(root.left, cur_sum, sum) + xxx(root.right, cur_sum, sum);
};

File: solutions/java/PathSumIII/PathSumIII.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode(int x) { val = x; }
 * }
 */
class Solution {
    public int pathSum(TreeNode root, int sum) {
        if (null == root) {
            return 0;
        }
        return cal(root, 0, sum) + pathSum(root.left, sum) + pathSum(root.right, sum);
    }

    private int cal(TreeNode root, int pre, int sum) {
        if (null == root) {
            return 0;
        }
        int cur_sum = pre + root.val;
        return (cur_sum == sum ? 1 : 0) + cal(root.left, cur_sum, sum) + cal(root.right, cur_sum, sum);
    }
}